### PR TITLE
Issue 402  gradient boundary

### PR DESCRIPTION
**Pull request by adler-j**
This adds boundary handling behavior to the `Gradient` family of operators.

* Added `periodic` boundary condition
* Added tests for many more boundary conditions using fixtures
* General cleanup of diff_ops tests
* Added all boundary conditions to `Gradient`, `Divergence` and `Laplacian`
* Added correct handling of `'constant'` boundary condition which causes the operator to become nonlinear.

### CODE REVIEW

File: odl/discr/diff_ops.py [hunk 1]
```diff
@@ -129,10 +143,37 @@ def _call(self, x, out=None):
...
+        else:
+            return self
+
     @property
     def adjoint(self):
         """Return the adjoint operator."""
-        return -self
+        if not self.is_linear:
+            raise ValueError('Operator with nonzero padding_value ({}) is not'
+                             ' linear and has no adjoint'
```
Comment (kohr-h) [inline odl/discr/diff_ops.py]: Dot at the end :-)

File: odl/discr/diff_ops.py [hunk 2]
```diff
@@ -161,6 +203,19 @@ def __init__(self, domain=None, range=None, method='forward'):
...
+            'constant' : Pads values outside the domain of ``f`` with a
+            constant value given by ``padding_value``.
+
+            'symmetric' : Pads with the reflection of the vector mirrored
+            along the edge of the array.
+
+            'periodic' : Pads with the values from the other side of the array.
+
+        padding_value : `float`, optional
+            If ``padding_method`` is 'constant' ``f`` assumes ``padding_value``
```
Comment (kohr-h) [inline odl/discr/diff_ops.py]: put a comma after `'constant'`

**Comment by adler-j** [comment]
Fixed the minor bugs by last commit, need final review.

File: test/discr/diff_ops_test.py [hunk 3]
```diff
@@ -27,16 +27,66 @@
...
+    dtype : dtype
+        The type of the output array
+
+    """
+    vec = [1]
+    vol = np.arange(lin_size, dtype=dtype)
+    for _ in range(ndim - 1):
+        vec.insert(0, lin_size)
+        vol = vol * vol.reshape(vec)
+    return vol
```
Comment (kohr-h) [inline test/discr/diff_ops_test.py]: We have this or something similar in `util.phantom`.

**Comment by kohr-h** [comment]
Check if we can use the utility in the test. Otherwise, I didn't spot anything unusual. Assuming the math is right, it's good to go.
